Implement more CheckedNumerics in RenderPassEncoder

darryl-apple opened the pull request:
#### fc13fa9398649ea9eb1034d84905b6db1ff65dc8
<pre>
Implement more CheckedNumerics in RenderPassEncoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=275618">https://bugs.webkit.org/show_bug.cgi?id=275618</a>
<a href="https://rdar.apple.com/130085918">rdar://130085918</a>

Reviewed by Mike Wyrzykowski.

Increase the usage of CheckedNumerics in RenderPassEncoder

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::runVertexBufferValidation):
(WebGPU::RenderPassEncoder::incrementDrawCount):
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::RenderPassEncoder::drawIndexed):
(WebGPU::RenderPassEncoder::drawIndexedIndirect):
(WebGPU::RenderPassEncoder::drawIndirect):
(WebGPU::RenderPassEncoder::errorValidatingPipeline const):
(WebGPU::RenderPassEncoder::setScissorRect):

Canonical link: <a href="https://commits.webkit.org/280201@main">https://commits.webkit.org/280201@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ed2a180663e2feb60dbc168dd8b6d28eccd445e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55921 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35245 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8389 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58917 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6353 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6549 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45026 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4387 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57950 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33135 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48215 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26164 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29918 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4496 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51885 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60507 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5934 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52456 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48284 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51969 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12408 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31074 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32158 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33240 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31906 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -386,7 +386,9 @@ static void setViewportMinMaxDepthIntoBuffer(auto& fragmentDynamicOffsets, float
         auto bufferSize = it->value.size;
-        if ((strideCount.value() - 1) * bufferData.stride + bufferData.lastStride > bufferSize) {
+        auto strideSize = checkedProduct<uint64_t>((strideCount.value() - 1), bufferData.stride);
+        auto requiredSize = checkedSum<uint64_t>(strideSize.value());
```
I think the above line should read:
```
auto requiredSize = checkedSum<uint64_t>(strideSize.value(),  bufferData.lastStride);
```
otherwise `bufferData.lastStride` is removed from the calculation, right?

darryl-apple replied:
This has been corrected.

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -450,8 +452,9 @@ static void setViewportMinMaxDepthIntoBuffer(auto& fragmentDynamicOffsets, float
-    m_drawCount += drawCalls;
-    if (m_drawCount > m_maxDrawCount)
+    auto checkedDrawCount = checkedSum<uint64_t>(m_drawCount, drawCalls);
+    m_drawCount = checkedDrawCount;
```
if checkedDrawCount overflows, I think this will crash because m_drawCount is a uint64_t and the definition of operator T in CheckedArithmetic.h is:
```
    operator T() const
    {
        if (UNLIKELY(this->hasOverflowed()))
            this->crash();
        return m_value;
    }
```

mwyrzykowski replied:
Proposed change, from:
```
m_drawCount = checkedDrawCount;
if (checkedDrawCount.hasOverflowed() || checkedDrawCount > m_maxDrawCount)
```
to:
```
m_drawCount = checkedDrawCount.hasOverflowed() ? (m_maxDrawCount + 1) : checkedDrawCount.value();
if (m_drawCount > m_maxDrawCount)
```

mwyrzykowski replied:
technically the +1 would overflow resulting in no limit if someone uses 2^64 - 1 for maxDrawCount but they will be waiting a long time for 2^64 -1 draw commands to occur, so I don't think this is an issue in practice

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -722,15 +731,21 @@ static void setViewportMinMaxDepthIntoBuffer(auto& fragmentDynamicOffsets, float
     }
 
-    if (firstIndexOffsetInBytes + indexCount * indexSizeInBytes > m_indexBufferSize) {
+    auto lastIndexOffset = checkedSum<NSUInteger>(firstIndexOffsetInBytes, indexCount * indexSizeInBytes);
```
I suppose there is a potential for overflow in `indexCount * indexSizeInBytes`, but since `indexSizeInBytes` is at most 4, you could write static_cast<uint64_t>(indexCount) * indexSizeInBytes but this is already a variable, `indexCountInBytes`. But on watchOS `indexCountInBytes` is only 32-bits, so it could overflow on watchOS on line 628, as `indexCountInBytes` is an NSUInteger instead of a uint64_t

Perhaps this isn't a concern as WebGPU does not support watchOS

mwyrzykowski reviewed (dismissed):
This is a great, thank you! I think there are a few small issues to address as commented, but please double check as I may have misread the logic. Otherwise looks good!

mwyrzykowski reviewed (commented):
This looks great, I'm going to apply the patch locally to double check but I didn't notice any issues

darryl-apple committed:
Implement more CheckedNumerics in RenderPassEncoder
https://bugs.webkit.org/show_bug.cgi?id=275618
rdar://130085918

Reviewed by Mike Wyrzykowski.

Increase the usage of CheckedNumerics in RenderPassEncoder

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::runVertexBufferValidation):
(WebGPU::RenderPassEncoder::incrementDrawCount):
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::RenderPassEncoder::drawIndexed):
(WebGPU::RenderPassEncoder::drawIndexedIndirect):
(WebGPU::RenderPassEncoder::drawIndirect):
(WebGPU::RenderPassEncoder::errorValidatingPipeline const):
(WebGPU::RenderPassEncoder::setScissorRect):

Canonical link: https://commits.webkit.org/280201@main
